### PR TITLE
Migrating `qiskit_algorithms`

### PR DESCRIPTION
**Pull request by edoaltamura**
### Summary
Qiskit algorithms, which Qiskit machine-learning (ML) depends on, is no longer officially supported as of https://github.com/qiskit-community/qiskit-algorithms/commit/4247d063c807af01a3ebae9018e189cebb80a112. To continue improving the subset of Algorithms features that are relevant to Qiskit machine-learning, it has been proposed to migrate those from Algorithms to ML.

This PR implements the merge of the subset of Qiskit Algorithms features needed for ML, as described in https://github.com/qiskit-community/qiskit-machine-learning/issues/811.

Closes https://github.com/qiskit-community/qiskit-machine-learning/issues/811.

* `qiskit_algorithms/gradients` -> `qiskit_machine_learning/gradients`
* `qiskit_algorithms/optimizers` -> `qiskit_machine_learning/optimizers`
* `qiskit_algorithms/state_fidelities` -> `qiskit_machine_learning/state_fidelities`
* `qiskit_algorithms/utils` partially merged with `qiskit_machine_learning/utils`

### Details and comments
The unit tests are also merged following the same structure as in the source directory.

* `qiskit_algorithms/tests/gradients` -> `qiskit_machine_learning/tests/gradients`
* `qiskit_algorithms/tests/optimizers` -> `qiskit_machine_learning/tests/optimizers`
* `qiskit_algorithms/tests/state_fidelities` -> `qiskit_machine_learning/tests/state_fidelities`
* `qiskit_algorithms/tests/utils` partially merged with `qiskit_machine_learning/tests/utils`

---
Once the merge is complete, the codebase will be ready for a targeted attempt at resolving:
* Compatibility with V2 primitives: https://github.com/qiskit-community/qiskit-machine-learning/issues/742 and https://github.com/qiskit-community/qiskit-machine-learning/issues/810, also in https://github.com/qiskit-community/qiskit-algorithms/issues/136
* Compliance with ISA standards: https://github.com/qiskit-community/qiskit-machine-learning/issues/786, also in https://github.com/qiskit-community/qiskit-algorithms/issues/164, https://github.com/qiskit-community/qiskit-algorithms/issues/165, https://github.com/qiskit-community/qiskit-finance/issues/338, https://github.com/qiskit-community/qiskit-nature/issues/1350 and https://github.com/Qiskit/qiskit-ibm-runtime/issues/1718

### CODE REVIEW

**Comment by woodsp-ibm** [comment]
I skimmed things very quickly but one thing it looks like you will need is to modify the main init docstring to include any new modules for the main table of contents there.

**Comment by woodsp-ibm** [comment]
For all the copyrights, you know you can do a `make copyright` locally and it will fix them all? I guess this needs to be done with the way they have been brought over right - I think its possible to bring them over with history, more effort/fuss of course, but I guess that's all there in algorithms should any need it for any reason.

**Comment by edoaltamura** [comment]
Thanks for the input, Steve. I fixed the copyright year and now going through the outstanding mypy errors. Then, I'll move over to the docstrings and release notes.

This should not be a breaking change because the objects generated from Algorithms itself would be equivalent to those now generated in ML. But it would be a good idea to produce warnings or messages at specific points in the logic to ensure users are aware of this change as we consolidate the library.

I can discuss the merge with Git history, however it might be more complex after changing the absolute/relative imports in some of the files.

**Comment by declanmillar** [comment]
Hey, @edoaltamura. Are you still working on this? It might be work marking it as a Draft until you think it's ready for review.

**Comment by woodsp-ibm** [comment]
I will note, that presently with the main branch of Qiskit, that is shortly to be released as 1.2.0 that there is an issue with the `gradients/utils.py` method `_make_lin_comb_gradient_circuit` that is copied over here in this PR - there is a "fix" in Qiskit Algorithms https://github.com/qiskit-community/qiskit-algorithms/pull/188 but I was holding off merging at present in case the underlying logic in Qiskit is remediated before release and this once again works. There is more info in that PR - also Julien did a PR to that PR where he changed the logic to use the DAG instead which works and whatever the Rust changes related issue that caused the existing logic to fail does not get triggered this way with the DAG.

**Comment by declanmillar** [comment]
Is it necessary for us to copy over all these files from algorithms when we're planning to cull most of them to reduce the maintenance load? I thought the intention described in #811 was to remove all but the SPSA gradient module. It seems counterproductive to fix issues in these files when they'll be removed soon.

**Comment by woodsp-ibm** [comment]
> Is it necessary for us to copy over all these files from algorithms when we're planning to cull 

I just saw that libcomb gradient had been copied here and noted the issue with present Qiskit main that fails the nightly CI unit tests in CI where it tests against Qiskit main branch.  In terms of the gradients it does seem to be a subset copied here of what is in algorithms, but I do recall the statement about SPSA one being the only reasonable choice for hardware.

**Comment by edoaltamura** [comment]
> > Regarding the spelling issues, to save time, we could do a union of the dictionary sets in qiskit-algorithms and qiskit-machine-learning. This will introduce some redundancies, but they should all be relevant words.
> 
> My recollection of doing something in the past was to remove the custom dict let it create a list of all misspelled words and then sort the resultant to create the custom dict. I think I did this to create a dict to start with where I looked through for evident mispellings and removed them - in this case I guess all mispellings are correct and we would want the entire list.

I have merged the two `.pylintdict` files, removed duplicates and sorted alphabetically again. This seems to have worked as the CI tests are passing for now.

**Comment by edoaltamura** [comment]
> > Is it necessary for us to copy over all these files from algorithms when we're planning to cull
> 
> I just saw that libcomb gradient had been copied here and noted the issue with present Qiskit main that fails the nightly CI unit tests in CI where it tests against Qiskit main branch. In terms of the gradients it does seem to be a subset copied here of what is in algorithms, but I do recall the statement about SPSA one being the only reasonable choice for hardware.

For the moment I have kept all `algorithms` sub-modules that are invoked in `machine-learning`, including `lin_comb` and `param_shift`. 
In some cases, like `estimator_qnn.py`, `ParamShiftEstimatorGradient` is used as a default but if we plan not to incorporate it we could make SPSA the new default at this level:
https://github.com/qiskit-community/qiskit-machine-learning/blob/89792d23bb633fb0fbf8fc4ad622af327337cb46/qiskit_machine_learning/neural_networks/estimator_qnn.py#L163-L164

https://github.com/qiskit-community/qiskit-machine-learning/blob/89792d23bb633fb0fbf8fc4ad622af327337cb46/qiskit_machine_learning/neural_networks/sampler_qnn.py#L181-L182

by replacing with
```python
gradient = SPSAEstimatorGradient(self.estimator, epsilon=1e-6)
```
and
```python
gradient = SPSASamplerGradient(self.sampler, epsilon=1e-6)
```
**Should we make SPSA the new default and only keep that?**
As a summary, this choice would drop
- `qiskit_machine_learning/gradients/lin_comb`: both linear combination of unitaries and Quantum Geometric Tensor (QGT)
- `qiskit_machine_learning/gradients/param_shift`
- `qiskit_machine_learning/gradients/base/base_qgt.py` and `qiskit_machine_learning/gradients/base/qgt_result.py` in the base directory.

**Comment by edoaltamura** [comment]
- The stable tutorial silencing worked. Before merging the PR I will open an issue to unblock those tests after version 0.8.0 is out.

- Removing the linear combination gradients is more tricky than I initially thought because it's tightly coupled with QGT routines and ultimately to Fischer Information (FI), which I was asked to keep. I think FI can be eventually decoupled from those in favour of other gradient methods, but I believe this goes beyond the scope of this PR. Therefore, I propose keeping the gradients as they currently are and issuing a deprecation down the line for those gradients that are not to be kept.

File: qiskit_machine_learning/optimizers/gradient_descent.py
Comment (declanmillar) [inline qiskit_machine_learning/optimizers/gradient_descent.py:183]:
```suggestion
        learning_rate: (
            float | list[float] | np.ndarray | Callable[[], Generator[float, None, None]]
        ) = 0.01,
```
Black formatted.

File: qiskit_machine_learning/optimizers/optimizer_utils/learning_rate.py
Comment (declanmillar) [inline qiskit_machine_learning/optimizers/optimizer_utils/learning_rate.py:35]:
```suggestion
        learning_rate: (
            float | list[float] | np.ndarray | Callable[[], Generator[float, None, None]]
        ),
```
Black formatted.

**Review by declanmillar — CHANGES_REQUESTED** [review]
Thanks for all your effort, Edo! I know a lot of these code quality and style tools can be a little overwhelming at first. This isn't a full review, but should fix those black errors.

File: releasenotes/notes/qiskit-algorithms-incorporated-421554a4ff547d0d.yaml
Comment (woodsp-ibm) [inline releasenotes/notes/qiskit-algorithms-incorporated-421554a4ff547d0d.yaml:16]: Do you want to noted that this was a subset of the gradients (reverse and finite-diff for instance were not copied over)

I do not think mention of unit tests is necessary for normal end-users.

File: releasenotes/notes/qiskit-algorithms-incorporated-421554a4ff547d0d.yaml
Comment (woodsp-ibm) [inline releasenotes/notes/qiskit-algorithms-incorporated-421554a4ff547d0d.yaml:23]: Issues are normally for known bugs in the code. Incorporating algorithms in my mind was to have those dependents here and supported (versus unsupported in algorithms). Basically come Aug 15th ML will no longer work with Runtime as V1 primitives will go away https://docs.quantum.ibm.com/announcements/product-updates/2024-07-15-v1-support-ending so it really does not facilitate anything in that regard.

File: releasenotes/notes/qiskit-algorithms-incorporated-421554a4ff547d0d.yaml
Comment (woodsp-ibm) [inline releasenotes/notes/qiskit-algorithms-incorporated-421554a4ff547d0d.yaml:52]: This is normally for deprecations that are added (stuff that when used would emit deprecation warnings). I think what this is getting at is already covered by upgrade section.

Maybe all these changes, being that they are just imports is straightforward enough that a migration guide is not needed - though possibly having a couple of concrete simple examples may be helpful to some

File: releasenotes/notes/qiskit-algorithms-incorporated-421554a4ff547d0d.yaml
Comment (woodsp-ibm) [inline releasenotes/notes/qiskit-algorithms-incorporated-421554a4ff547d0d.yaml:42]: As imports and sphinx refs use `.` would it be preferable here to show them that way e.g.

> Change `qiskit_algorithms.gradients` to `qiskit_machine_learning.gradients`

File: requirements.txt
Comment (woodsp-ibm) [inline requirements.txt:1]: >=0.44, I wonder how true this is still. Maybe this should be bumped to 1.0.0 as a min? - though I guess its probably still good with 0.46.

File: releasenotes/notes/qiskit-algorithms-incorporated-421554a4ff547d0d.yaml
Comment (edoaltamura) [reply releasenotes/notes/qiskit-algorithms-incorporated-421554a4ff547d0d.yaml:23]: Removed the issues section.

File: releasenotes/notes/qiskit-algorithms-incorporated-421554a4ff547d0d.yaml
Comment (edoaltamura) [reply releasenotes/notes/qiskit-algorithms-incorporated-421554a4ff547d0d.yaml:42]: Changed using `.`

File: releasenotes/notes/qiskit-algorithms-incorporated-421554a4ff547d0d.yaml
Comment (edoaltamura) [reply releasenotes/notes/qiskit-algorithms-incorporated-421554a4ff547d0d.yaml]: Added four examples.

File: requirements.txt
Comment (edoaltamura) [reply requirements.txt]: Set 1.0 as a minimum.

File: releasenotes/notes/qiskit-algorithms-incorporated-421554a4ff547d0d.yaml
Comment (declanmillar) [inline releasenotes/notes/qiskit-algorithms-incorporated-421554a4ff547d0d.yaml:102]: This looks fine in the rendered docs, but perhaps we should rewrap the lines over 100 characters long in the source for consistency with the rest of the repo?

**Review by declanmillar — APPROVED** [review]
I left a minor style suggestion. Otherwise, LGTM.

File: releasenotes/notes/qiskit-algorithms-incorporated-421554a4ff547d0d.yaml
Comment (edoaltamura) [reply releasenotes/notes/qiskit-algorithms-incorporated-421554a4ff547d0d.yaml]: Applied the line breaks at 100 chars.

File: releasenotes/notes/qiskit-algorithms-incorporated-421554a4ff547d0d.yaml
Comment (declanmillar) [inline releasenotes/notes/qiskit-algorithms-incorporated-421554a4ff547d0d.yaml:114]: Rather than truncating the line, you can use automated scripts to cleanly rewrap the line. If you use VSCode, for example, you can use `Rewrap Comment / Text` on highlighted paragraphs.

Comment (edoaltamura) [reply releasenotes/notes/qiskit-algorithms-incorporated-421554a4ff547d0d.yaml]: Adjusted the line breaks

**Comment by declanmillar** [comment]
LGTM. Thanks again :)

**Review by OkuyanBoga — APPROVED** [review]
I think it is good to go. Thanks @edoaltamura

**Pull request merged**
